category_theory/limits/preserves.lean

```diff
@@ -0,0 +1,155 @@
+-- Copyright (c) 2018 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Scott Morrison
+
+import category_theory.whiskering
+import category_theory.limits.limits
+
+open category_theory
+
+namespace category_theory.limits
+
+universes u₁ u₂ v
+
+variables {C : Type u₁} [𝒞 : category.{u₁ v} C]
+variables {D : Type u₂} [𝒟 : category.{u₂ v} D]
+include 𝒞 𝒟
+
+variables {J : Type v} [small_category J] {K : J ⥤ C}
+
+-- FIXME Reid's suggestion using forall, also for all variants of has_limits
+-- TODO rename to `continuous`?
+-- TODO show `functor.id` preserves limits?
+
+class preserves_limit (K : J ⥤ C) (F : C ⥤ D) :=
+(preserves : Π {c : cone K}, is_limit c → is_limit (F.map_cone c))
+class preserves_colimit (K : J ⥤ C) (F : C ⥤ D) :=
+(preserves : Π {c : cocone K}, is_colimit c → is_colimit (F.map_cocone c))
+
+class preserves_limits_of_shape (J : Type v) [small_category J] (F : C ⥤ D) :=
+(preserves : Π {K : J ⥤ C} {c : cone K}, is_limit c → is_limit (F.map_cone c))
+class preserves_colimits_of_shape (J : Type v) [small_category J] (F : C ⥤ D) :=
+(preserves : Π {K : J ⥤ C} {c : cocone K}, is_colimit c → is_colimit (F.map_cocone c))
+
+class preserves_limits (F : C ⥤ D) :=
+(preserves : Π {J : Type v} [small_category J] {K : J ⥤ C} {c : cone K},
+  is_limit c → is_limit (F.map_cone c))
+class preserves_colimits (F : C ⥤ D) :=
+(preserves : Π {J : Type v} [small_category J] {K : J ⥤ C} {c : cocone K},
+  is_colimit c → is_colimit (F.map_cocone c))
+
+instance preserves_limit_of_preserves_limits_of_shape
+  (K : J ⥤ C) (F : C ⥤ D) [preserves_limits_of_shape J F] :
+  preserves_limit K F :=
+{ preserves := λ _, preserves_limits_of_shape.preserves F }
+
+instance preserves_limits_of_shape_of_preserves_limit (F : C ⥤ D) [preserves_limits F] :
+  preserves_limits_of_shape J F :=
+{ preserves := λ _ _, preserves_limits.preserves F }
+
+-- def preserves_limit.is_limit {F : C ⥤ D} [preserves_limit K F]
```

Comment:
comment

Reply:
I've added a note `-- TODO` beforehand.

I hope this is okay -- I think it's helpful to leave in a few pointers to extra things people can add when needed or wanted, but I'm happy to remove these as well, or perhaps even get around to filling them in...